packages/beacon-state-transition/src/allForks/util/weakSubjectivity.ts

```diff
@@ -15,3 +30,62 @@ export function getLatestWeakSubjectivityCheckpointEpoch(
   const valCount = state.epochCtx.currentShuffling.activeIndices.length;
   return getWeakSubjectivityCheckpointEpoch(config, state.finalizedCheckpoint.epoch, valCount, safetyDecay);
 }
+
+/**
+  Returns the weak subjectivity period for the current ``state``. 
+    This computation takes into account the effect of:
+      - validator set churn (bounded by ``get_validator_churn_limit()`` per epoch), and 
+      - validator balance top-ups (bounded by ``MAX_DEPOSITS * SLOTS_PER_EPOCH`` per epoch).
+    A detailed calculation can be found at:
+    https://github.com/runtimeverification/beacon-chain-verification/blob/master/weak-subjectivity/weak-subjectivity-analysis.pdf
+ */
+export function computeWeakSubjectivityPeriod(config: IBeaconConfig, state: allForks.BeaconState): Uint64 {
+  let wsPeriod = config.params.MIN_VALIDATOR_WITHDRAWABILITY_DELAY;
+  const N = getActiveValidatorIndices(state, getCurrentEpoch(config, state)).length;
+  const t = Math.floor(Number(getTotalActiveBalance(config, state)) / N / Number(ETH_TO_GWEI));
+  const T = Math.floor(Number(config.params.MAX_EFFECTIVE_BALANCE) / Number(ETH_TO_GWEI));
+  const delta = getValidatorChurnLimit(config, state);
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  const Delta = config.params.MAX_DEPOSITS * config.params.SLOTS_PER_EPOCH;
+  const D = Number(SAFETY_DECAY);
+
+  if (T * (200 + 3 * D) < t * (200 + 12 * D)) {
+    const epochsForValidatorSetChurn = Math.floor(
+      (N * (t * (200 + 12 * D) - T * (200 + 3 * D))) / (600 * delta * (2 * t + T))
+    );
+    const epochsForBalanceTopUps = Math.floor((N * (200 + 3 * D)) / (600 * Delta));
+    wsPeriod +=
+      epochsForValidatorSetChurn > epochsForBalanceTopUps ? epochsForValidatorSetChurn : epochsForBalanceTopUps;
+  } else {
+    wsPeriod += Math.floor((3 * N * D * t) / (200 * Delta * (T - t)));
+  }
+  return BigInt(wsPeriod);
+}
+
+function getLatestBlockRoot(config: IBeaconConfig, state: allForks.BeaconState): Root {
+  const header = config.types.phase0.BeaconBlockHeader.clone(state.latestBlockHeader);
+  if (config.types.Root.equals(header.stateRoot, ZERO_HASH)) {
+    header.stateRoot = config.getForkTypes(state.slot).BeaconState.hashTreeRoot(state);
+  }
+  return config.types.phase0.BeaconBlockHeader.hashTreeRoot(header);
+}
+
+export function isWithinWeakSubjectivityPeriod(
+  config: IBeaconConfig,
+  genesisTime: number,
+  wsState: allForks.BeaconState,
+  wsCheckpoint: Checkpoint
+): boolean {
+  const blockRoot = getLatestBlockRoot(config, wsState);
+  if (!config.types.Root.equals(blockRoot, wsCheckpoint.root)) {
+    throw new Error(
+      `Roots do not match.  expected=${toHexString(wsCheckpoint.root)}, actual=${toHexString(blockRoot)}`
+    );
+  }
+  assert.equal(computeEpochAtSlot(config, wsState.slot), wsCheckpoint.epoch);
```

Comment:
you should probably add third param message

Reply:
Yes please :pray: